### PR TITLE
feat(ir): add tile.mscatter op for per-element scatter-store to GM

### LINKED ISSUE #921: [New Op] Add op for pto.mscatter instruction
### Operation Level

Block-level

### Proposed Name & Signature

TBD — needs investigation into the `pto.mscatter` instruction semantics.

Possible tile-level op name following existing conventions:

```python
pl.mscatter(src: Tile, mask: Tile) -> Tile
```

For reference, the adjacent masked gather/scatter codegen entries exist in `pto_ops_common.cpp`:
- `tile.mgather` → `pto.tmgather` (2 args)
- `tile.mscatter` → `pto.tmscatter` (2 args)

Note: `pto.mscatter` (without `t` prefix) may be a distinct instruction from `pto.tmscatter`. The exact relationship and semantics need to be clarified.

### Semantics Description

Masked scatter operation corresponding to the `pto.mscatter` PTO instruction.

Currently, there is no PyPTO op that maps to `pto.mscatter`. The implementation should include:
1. C++ op definition in `src/ir/op/tile_ops/`
2. Python IR API in `python/pypto/ir/op/tile_ops.py`
3. Python DSL API in `python/pypto/language/op/tile_ops.py`
4. Codegen mapping in `pto_ops_common.cpp` (if `pto.mscatter` differs from `pto.tmscatter`)
5. Type stubs in `python/pypto/pypto_core/*.pyi`
6. Unit tests

### Example Usage

```python
@pl.program
class Example:
    @pl.function
    def main(self, ...):
        result = pl.mscatter(src, mask)
        ...
```

(Exact usage depends on finalized semantics.)

### Hardware Support Considerations

- Maps to PTO instruction `pto.mscatter`
- Relationship to existing `pto.tmscatter` (mapped from `tile.mscatter`) needs clarification
- Note: `tile.mgather` / `tile.mscatter` codegen entries already exist but have no corresponding op definitions or Python APIs either

### Motivation / Use Case

Complete the PyPTO op coverage for PTO scatter instructions. Currently `pto.mscatter` has no corresponding op in the PyPTO IR, preventing users from utilizing this hardware capability.

### Additional Context

Related: #677, #920

Existing scatter/gather landscape in `pto_ops_common.cpp`:
- `tile.gather` → `pto.tgather` (3 args) — has full Python API
- `tile.gatherb` → `pto.tgatherb` (2 args) — codegen mapping only
- `tile.scatter` → `pto.tscatter` (2 args) — codegen mapping only
- `tile.mgather` → `pto.tmgather` (2 args) — codegen mapping only, no op def
- `tile.mscatter` → `pto.tmscatter` (2 args) — codegen mapping only, no op def
- `tile.scatter_update` — has Python API, tensor→tile conversion, but no codegen mapping (#920)

### PR DESCRIPTION
**Pull request by Little-oil**
## Summary

Add `tile.mscatter` operation that maps to the PTOAS `pto.mscatter` instruction for per-element scatter-store from UB tile to GM tensor:

```
output_tensor[idx[i, j]] = src[i, j]
```

### Changes

- **C++ op registration** (`src/ir/op/tile_ops/memory.cpp`): `tile.mscatter` with type deduction validating src (FP16/FP32/INT16/INT32), idx (INT32, same shape as src), and output_tensor (TensorType, non-scalar, same dtype as src)
- **PTO codegen** (`src/backend/common/pto_ops_common.cpp`): Emits `pto.partition_view` + `pto.mscatter ins(src, idx) outs(pview)` with row_major layout constraints on inputs
- **Python IR wrapper** (`python/pypto/ir/op/tile_ops.py`): `tile.mscatter(src, idx, output_tensor)`
- **Python DSL wrapper** (`python/pypto/language/op/tile_ops.py`): `pl.mscatter(src_tile, idx_tile, out_tensor)` exported at top-level `pl` namespace and added to `__all__`
- **Unit tests** (`tests/ut/ir/operators/test_tile_ops.py`): 9 tests covering basic usage (FP32/FP16), error paths (wrong dtype, rank mismatch, shape mismatch, scalar output, arg count)
- **ST runtime tests** (`tests/st/runtime/test_mscatter.py`): 11 tests covering FP32/FP16/INT32, 8x32/16x64 shapes, sequential/reversed/random/strided index patterns — all 11 passed on NPU device
- **CI updates** (`.github/workflows/ci.yml`):
  - Bumped `--pto-isa-commit` to `ed0b4643` (includes mscatter CPU simulator fix `12f663c3`)
  - Bumped PTOAS to v0.26 for both system-tests (aarch64) and a5sim (x86_64) jobs

### Platform Support

`pto.mscatter` is currently only supported on A5 (Ascend 950). A3 (Ascend 910B) does not yet support this instruction in PTOAS, so ST tests are marked with `@pytest.mark.a5` and are not executed on 910B.

## Testing

- [x] Unit tests pass (9/9)
- [x] ST runtime tests pass on NPU device (11/11, A5 only)
- [x] CI passes (all jobs green)
- [x] Pre-commit hooks pass (clang-format, cpplint, ruff, pyright)

Fixes #921

### CODE REVIEW

File: src/ir/op/tile_ops/memory.cpp [hunk 1]
```diff
@@ -535,6 +535,65 @@ REGISTER_OP("tile.store")
...
+  // Second arg: idx tile (INT32, same rank as src)
+  auto idx_type = As<TileType>(args[1]->GetType());
+  CHECK(idx_type) << "The operator " << op_name << " requires second argument to be a TileType, but got "
+                  << args[1]->GetType()->TypeName();
+  CHECK(idx_type->dtype_ == DataType::INT32)
+      << "The operator " << op_name << " requires idx dtype to be INT32, but got "
+      << idx_type->dtype_.ToString();
+  CHECK(idx_type->shape_.size() == src_type->shape_.size())
+      << "The operator " << op_name << " requires idx rank to match src rank (" << src_type->shape_.size()
+      << "), but got " << idx_type->shape_.size();
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tile_ops/memory.cpp:587]: Fixed in d6a0a5e7 — added per-dimension shape equality validation between idx and src tiles.

File: src/ir/op/tile_ops/memory.cpp [hunk 2]
```diff
@@ -535,6 +535,65 @@ REGISTER_OP("tile.store")
...
+
+  // Third arg: output tensor (same dtype as src)
+  auto tensor_type = As<TensorType>(args[2]->GetType());
+  CHECK(tensor_type) << "The operator " << op_name << " requires third argument to be a TensorType, but got "
+                     << args[2]->GetType()->TypeName();
+  CHECK(tensor_type->dtype_ == src_type->dtype_)
+      << "The operator " << op_name << " requires output_tensor dtype (" << tensor_type->dtype_.ToString()
+      << ") to match src dtype (" << src_type->dtype_.ToString() << ")";
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tile_ops/memory.cpp:607]: Fixed in d6a0a5e7 — added CHECK that output_tensor has at least 1 dimension.

File: python/pypto/language/__init__.py [hunk 3]
```diff
@@ -130,6 +130,9 @@ def scalar_func(x: pl.Scalar[pl.FP32]) -> pl.Scalar[pl.FP32]:
     xor,
     xors,
 )
+from .op.tile_ops import (
+    mscatter as mscatter,
+)
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply python/pypto/language/__init__.py:139]: mscatter is a tile-level operation (UB→GM scatter-store), not a system/cross-core op. Importing from tile_ops is correct and follows the same pattern as load/store.

File: python/pypto/language/__init__.py [hunk 4]
```diff
@@ -133,6 +133,9 @@ def scalar_func(x: pl.Scalar[pl.FP32]) -> pl.Scalar[pl.FP32]:
...
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply python/pypto/language/__init__.py:139]: Fixed in d6a0a5e7 — added mscatter to __all__.

File: src/backend/common/pto_ops_common.cpp [hunk 5]
```diff
@@ -713,6 +713,84 @@ static std::string MakeTileStoreCodegenPTO(const CallPtr& op, codegen::CodegenBa
...
+      partition_line << codegen.GetIndexConstant(c->value_);
+      partition_type += std::to_string(c->value_);
+    } else {
+      partition_line << codegen.GetExprAsCode(tensor_type->shape_[i]);
+      partition_type += "?";
+    }
+  }
+  partition_line << "]";
+  partition_type += "x" + dtype_str + ">";
+  partition_line << " : " << tensor_view_type << " -> " << partition_type;
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/backend/common/pto_ops_common.cpp:796]: mscatter uses all-zero offsets and full tensor shape for the partition_view — DN swap on zero offsets is a no-op, and GetOrCreateTensorView already handles the physical layout. In practice, mscatter targets 1D output tensors where DN is not applicable.

File: tests/st/runtime/test_mscatter.py [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+            TensorSpec("output", [512], DataType.FP32, is_output=True),
+        ]
+
+    def get_program(self) -> Any:
+        return MscatterFP32_8x32_LargeOutputProgram
+
+    def compute_expected(self, tensors, params=None):
+        out = torch.zeros(512, dtype=torch.float32)
+        out[tensors["idx_tensor"].flatten().long()] = tensors["src_tensor"].flatten()
+        tensors["output"][:] = out
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply tests/st/runtime/test_mscatter.py:350]: Fixed in d6a0a5e7 — added init_value=torch.zeros for the sparse scatter (strided) test case output buffer.

**Review by lyfne123 — APPROVED** [review]
